rct229/rules/section22/section22rule8.py

```diff
@@ -0,0 +1,126 @@
+            each_rule=Section22Rule8.PumpRule(),
+            index_rmr="baseline",
+            id="22-8",
+            description="For Baseline chilled water system with cooling capacity of 300 tons or more, the secondary pump shall be modeled with variable-speed drives.",
+            rmr_context="ruleset_model_instances/0",
+            list_path="pumps[*]",
+        )
+
+    def is_applicable(self, context, data=None):
+        rmi_b = context.baseline
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+        # create a list contains all HVAC systems that are modeled in the rmi_b
+        available_type_lists = [
+            hvac_type
+            for hvac_type in baseline_system_types_dict.keys()
+            if len(baseline_system_types_dict[hvac_type]) > 0
+        ]
+
+        primary_secondary_loop_dictionary = get_primary_secondary_loops_dict(rmi_b)
+        return (
+            any(
+                [
+                    available_type in APPLICABLE_SYS_TYPES
+                    for available_type in available_type_lists
+                ]
+            )
+            and primary_secondary_loop_dictionary
+        )
+
+    def create_data(self, context, data):
+        rmi_b = context.baseline
+        loop_pump_dictionary = {
+            pump["loop_or_piping"]: pump for pump in find_all("pumps[*]", rmi_b)
+        }
+
+        chw_loop_capacity_dict = {}
+        for chiller in find_all("chillers[*]", rmi_b):
+            if chiller["cooling_loop"] not in chw_loop_capacity_dict.keys():
+                chw_loop_capacity_dict[chiller["cooling_loop"]] = 0.0 * ureg("W")
+                chw_loop_capacity_dict[chiller["cooling_loop"]] += chiller[
+                    "rated_capacity"
+                ]
+
+        primary_secondary_loop_dictionary = get_primary_secondary_loops_dict(rmi_b)
+
+        child_loop_speed_control_dict = {}
+        for loop_b_id in primary_secondary_loop_dictionary.keys():
+            if (
+                chw_loop_capacity_dict[loop_b_id]
+                >= MIN_CHW_PRIMARY_LOOP_COOLING_CAPACITY
+            ):
+                for child_loop in find_one_with_field_value(
+                    "fluid_loops[*]", "id", loop_b_id, rmi_b
+                )["child_loops"]:
+                    child_loop_speed_control_dict[
+                        child_loop["id"]
+                    ] = loop_pump_dictionary[child_loop["id"]]["speed_control"]
+
+        return {
+            "loop_pump_dictionary": loop_pump_dictionary,
+            "chw_loop_capacity_dict": chw_loop_capacity_dict,
+            "child_loop_speed_control_dict": child_loop_speed_control_dict,
+        }
+
+    def list_filter(self, context_item, data):
+        pump_b = context_item.baseline
+        child_loop_speed_control_dict = data["child_loop_speed_control_dict"]
+
+        return pump_b["loop_or_piping"] in child_loop_speed_control_dict.keys()
+
+    class PumpRule(RuleDefinitionBase):
```

Comment:
I think this should be restructured to: ChillerFluidLoop class -> SecondaryChildLoop class.
The loop_pump_dictionary is called in the `calc_vals` in the SecondaryChildLoop class.
From report point of view, we want to include the `id` of the secondary loop (this is included in the SecondaryChildLoop class).

In the Secondary loop class, we can get the pump from the `loop_pump_dictionary` and take out the `speed_control` for rule assertion.

The RDS is a bit twist but I think following FluidLoop path is cleaner.

Reply:
The code was restructured to have ```ChillerFluidLoop ``` and ```SecondaryChildLoop ``` classes. Thanks for letting me know about this!